File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/ir/InlineVisitor.java
package ro.pub.cs.diploma.ir;

import com.intellij.psi.PsiCodeBlock;
import com.intellij.psi.PsiElementFactory;
import com.intellij.psi.PsiStatement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;

public class InlineVisitor implements Visitor {
  @NotNull private final PsiElementFactory factory;
  @NotNull private final String blockSet;
  @NotNull private final PsiCodeBlock block;

  @NotNull private PsiCodeBlock currentBlock;

  @NotNull
  private PsiCodeBlock newBlock() {
    return factory.createCodeBlock();
  }

  private void addStatement(@NotNull final String text) {
    currentBlock.add(factory.createStatementFromText(text, null));
  }

  private void addStatement(@NotNull final WrapperStatement statement) {
    currentBlock.add(statement.getStatement());
  }

  private void addStatement(@NotNull final PsiStatement statement) {
    currentBlock.add(statement);
  }

  private void addBlockSet(@NotNull final String val) {
    addStatement(blockSet + val + ";");
    addStatement("break;");
  }

  public InlineVisitor(@NotNull final PsiElementFactory factory, @NotNull final String frameVarName, @NotNull final String blockFieldName) {
    this.factory = factory;
    blockSet = frameVarName + "." + blockFieldName + " = ";
    block = newBlock();

    currentBlock = block;
  }

  @NotNull
  public PsiCodeBlock getBlock() {
    return block;
  }

  @Override
  public void visit(@NotNull final Block block) {
    block.getStatements().forEach(statement -> statement.accept(this));
  }

  @Nullable
  private PsiCodeBlock inline(@NotNull final Block block) {
    PsiCodeBlock psiBlock = null;
    if (block.isInline()) {
      psiBlock = newBlock();
      final PsiCodeBlock oldCurrentBlock = currentBlock;
      currentBlock = psiBlock;
      block.accept(this);
      currentBlock = oldCurrentBlock;
    }
    return psiBlock;
  }

  @NotNull
  private PsiCodeBlock getConcreteBlock(@NotNull final Block block, @Nullable final PsiCodeBlock psiBlock) {
    PsiCodeBlock concretePsiBlock;
    if (psiBlock != null) {
      concretePsiBlock = psiBlock;
    }
    else {
      concretePsiBlock = newBlock();
      final PsiCodeBlock oldCurrentBlock = currentBlock;
      currentBlock = concretePsiBlock;
      addBlockSet(String.valueOf(block.getId()));
      currentBlock = oldCurrentBlock;
    }
    return concretePsiBlock;
  }

  @Override
  public void visit(@NotNull ConditionalJumpStatement conditionalJumpStatement) {
    final Block thenBlock = conditionalJumpStatement.getThenBlock();
    final PsiCodeBlock thenPsiBlock = inline(thenBlock);

    final Block elseBlock = conditionalJumpStatement.getElseBlock();
    final PsiCodeBlock elsePsiBlock = inline(elseBlock);

    final String conditionText = conditionalJumpStatement.getCondition().getText();

    if (thenPsiBlock == null && elsePsiBlock == null) {
      addBlockSet(conditionText + " ? " + thenBlock.getId() + " : " + elseBlock.getId());
      return;
    }

    final PsiCodeBlock concreteThenPsiBlock = getConcreteBlock(thenBlock, thenPsiBlock);
    final PsiCodeBlock concreteElsePsiBlock = getConcreteBlock(elseBlock, elsePsiBlock);

    addStatement("if (" + conditionText + ")" + concreteThenPsiBlock.getText() + " else " + concreteElsePsiBlock.getText());
  }

  @Override
  public void visit(@NotNull final NormalStatement normalStatement) {
    addStatement(normalStatement);
  }

  @Override
  public void visit(@NotNull final ReturnStatement returnStatement) {
    addStatement(returnStatement);
  }

  @Override
  public void visit(@NotNull final UnconditionalJumpStatement unconditionalJumpStatement) {
    final Block block = unconditionalJumpStatement.getBlock();

    final PsiCodeBlock psiBlock = inline(block);

    if (psiBlock != null) {
      Arrays.stream(psiBlock.getStatements()).forEach(this::addStatement);
      return;
    }

    addBlockSet(String.valueOf(block.getId()));
  }
}

File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/ir/InlineVisitor.kt
package ro.pub.cs.diploma.ir

import com.intellij.psi.PsiCodeBlock
import com.intellij.psi.PsiElementFactory
import ro.pub.cs.diploma.NameManager
import ro.pub.cs.diploma.statement

class InlineVisitor(private val factory: PsiElementFactory, nameManager: NameManager) : Visitor {
  private val blockSet: String = "${nameManager.frameVarName}.${nameManager.blockFieldName} = "
  val block: PsiCodeBlock = factory.createCodeBlock()
  private var currentBlock: PsiCodeBlock = block

  private fun addStatement(text: String) {
    currentBlock.add(factory.statement(text))
  }

  private fun addStatement(statement: WrapperStatement) {
    currentBlock.add(statement.statement)
  }

  private fun addBreak() {
    addStatement("break;")
  }

  private fun addBlockSet(value: String) {
    addStatement("$blockSet$value;")
    addBreak()
  }

  override fun visit(block: Block) {
    block.statements.forEach { statement -> statement.accept(this) }
  }

  private fun inline(block: Block): PsiCodeBlock? {
    var psiBlock: PsiCodeBlock? = null
    if (block.isInlinable) {
      psiBlock = factory.createCodeBlock()
      val oldCurrentBlock = currentBlock
      currentBlock = psiBlock
      block.accept(this)
      currentBlock = oldCurrentBlock
    }
    return psiBlock
  }

  private fun getConcreteBlock(block: Block, psiBlock: PsiCodeBlock?): PsiCodeBlock {
    val concretePsiBlock: PsiCodeBlock
    if (psiBlock != null) {
      concretePsiBlock = psiBlock
    } else {
      concretePsiBlock = factory.createCodeBlock()
      val oldCurrentBlock = currentBlock
      currentBlock = concretePsiBlock
      addBlockSet(block.id.toString())
      currentBlock = oldCurrentBlock
    }
    return concretePsiBlock
  }

  override fun visit(conditionalJumpStatement: ConditionalJumpStatement) {
    val thenBlock = conditionalJumpStatement.thenBlock
    val thenPsiBlock = inline(thenBlock)

    val elseBlock = conditionalJumpStatement.elseBlock
    val elsePsiBlock = inline(elseBlock)

    val conditionText = conditionalJumpStatement.condition.text

    if (thenPsiBlock == null && elsePsiBlock == null) {
      addBlockSet("$conditionText ? ${thenBlock.id} : ${elseBlock.id}")
      return
    }

    val concreteThenPsiBlock = getConcreteBlock(thenBlock, thenPsiBlock)
    val concreteElsePsiBlock = getConcreteBlock(elseBlock, elsePsiBlock)

    addStatement("if ($conditionText) ${concreteThenPsiBlock.text} else ${concreteElsePsiBlock.text}")
  }

  override fun visit(normalStatement: NormalStatement) {
    addStatement(normalStatement)
  }

  override fun visit(returnStatement: ReturnStatement) {
    addStatement(returnStatement)
  }

  override fun visit(unconditionalJumpStatement: UnconditionalJumpStatement) {
    val block = unconditionalJumpStatement.block
    val psiBlock = inline(block)

    if (psiBlock != null) {
      psiBlock.statements.forEach { currentBlock.add(it) }
      return
    }

    addBlockSet(block.id.toString())
  }

  override fun visit(switchStatement: SwitchStatement) {
    val statements = switchStatement.statements
    val psiBlock = factory.createCodeBlock()
    val oldCurrentBlock = currentBlock
    currentBlock = psiBlock
    for (statement in statements) {
      statement.accept(this)
    }
    currentBlock = oldCurrentBlock

    addStatement("switch (${switchStatement.expression.text}) ${psiBlock.text}")
    addBreak()
  }
}
